File: lib/messages/Messages.ts
import { v4 as uuidv4 } from 'uuid';

// Header
export interface MessageHeader {
    requestId: string;
    messagePurpose: string;
    messageType?: string;
    version: number;
}

// Body
export interface MessageBody {
    [key: string]: any;
}

// Message
export class Message {
    // Properties
    header: MessageHeader = {
        requestId: uuidv4(),
        messagePurpose: "",
        messageType: "",
        version: 1
    };
    body: MessageBody = {};

    // Methods
    getUUID(): string {
        return this.header.requestId;
    }
}

// ErrorMessage
interface ErrorMessageBody extends MessageBody {
    statusCode: number;
    statusMessage: string;
}

export class ErrorMessage extends Message {
    body: ErrorMessageBody;

    // Constructor
    constructor(statusCode: number, statusMessage: string) {
        super();
        this.body = {
            statusCode: statusCode,
            statusMessage: statusMessage
        };
        this.header.messagePurpose = "error";
    }
}

File: src/messages/messages.go
package messages

import "github.com/google/uuid"

type Message struct {
	Server string        `json:"server"`
	Header MessageHeader `json:"header"`
	Body   interface{}   `json:"body"`
}

// MessageHeader - The header of a message
type MessageHeader struct {
	RequestId      uuid.UUID `json:"requestId"`
	MessagePurpose string    `json:"messagePurpose"`
	MessageType    string    `json:"messageType"`
	Version        int       `json:"version"`
}

// header.messagePurpose = "error"
// ErrorMessageBody - The body of an error message
type ErrorMessageBody struct {
	StatusCode    int    `json:"statusCode"`
	StatusMessage string `json:"statusMessage"`
}

// EventBody - The body of an event message
type EventBody struct {
	EventName string `json:"eventName"`
}

// NewMessage - Create a new message
func NewEventMessage(eventName string) *Message {
	return &Message{
		Header: MessageHeader{
			MessagePurpose: "",
			MessageType:    "commandResponse",
			Version:        1,
		},
		Body: EventBody{
			EventName: eventName,
		},
	}
}

// NewEventSubscribeMessage - Create a new event subscribe message
func NewEventSubscribeMessage(eventName string) *Message {
	message := NewEventMessage(eventName)
	message.Header.MessagePurpose = "subscribe"
	return message
}

// NewEventUnsubscribeMessage - Create a new event unsubscribe message
func NewEventUnsubscribeMessage(eventName string) *Message {
	message := NewEventMessage(eventName)
	message.Header.MessagePurpose = "unsubscribe"
	return message
}
